[MM-44084] Feature: Top threads insights

plant99 opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does, as well as QA test steps (if applicable and if not already added to the Jira ticket).
-->
Adds new endpoint to get top threads for a team and user.

This adds two new endpoints:
GET /api/v4/teams/:team_id/top/threads and GET /api/v4/users/me/top/threads.

Top threads for a user can also be scoped to a team by adding the team_id as a parameter to the request.

Sample queries:

```
curl -L -X GET 'http://localhost:8065/api/v4/teams/gb8ryzwgnir9xgndyfmhai7sjw/top/threads?time_range=28_day&page=0&per_page=5'\
    -H 'X-Requested-With: XMLHttpRequest'\
    -H 'Authorization: Bearer <token>'


curl -L -X GET 'http://localhost:8065/api/v4/users/me/top/threads?time_range=28_day&page=0&per_page=5&team_id=gb8ryzwgnir9xgndyfmhai7sjw'\
    -H 'X-Requested-With: XMLHttpRequest'\
    -H 'Authorization: Bearer <token>'
```


#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
JIRA: https://mattermost.atlassian.net/browse/MM-44084
#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
Added new API endpoints GET /api/v4/teams/:team_id/top/threads, GET /api/v4/users/me/top/threads
```

plant99 commented:
> Code looks great! We are just missing some information in the response. I need some more information in order to create the widget:
> 
> 1. User ID, and user.last_picture_update for the profile picture,
> 2. user.first_name & user.last_name for the name portion
> 3. Channel display name and the channel name (for the link)
> 4. The post message
> 5. particapant information
> 
> Check out the figma [here](https://www.figma.com/file/fKyLulu6MUjQzQa9pBFFrC/Insights-and-Analytics?node-id=586%3A117159) and you can see the threads widget and modal

@BenCookie95 , I've updated response to have the things you mentioned.

BenCookie95 commented:
> > Code looks great! We are just missing some information in the response. I need some more information in order to create the widget:
> > 
> > 1. User ID, and user.last_picture_update for the profile picture,
> > 2. user.first_name & user.last_name for the name portion
> > 3. Channel display name and the channel name (for the link)
> > 4. The post message
> > 5. particapant information
> > 
> > Check out the figma [here](https://www.figma.com/file/fKyLulu6MUjQzQa9pBFFrC/Insights-and-Analytics?node-id=586%3A117159) and you can see the threads widget and modal
> 
> @BenCookie95 , I've updated response to have the things you mentioned.

Looks great aside from some things I noticed while adding the frontend:

1. I noticed that there will also be instances where the message might be a file or an image. This means that we would have to include the metadata in the response
2. The response still includes threads even if the parent message is deleted
3. The response payload is missing the author details

@mkraft Do you think it's possible for us to handle file/image attachments? Or will there be too many joins?

plant99 commented:
> > > Code looks great! We are just missing some information in the response. I need some more information in order to create the widget:
> > > 
> > > 1. User ID, and user.last_picture_update for the profile picture,
> > > 2. user.first_name & user.last_name for the name portion
> > > 3. Channel display name and the channel name (for the link)
> > > 4. The post message
> > > 5. particapant information
> > > 
> > > Check out the figma [here](https://www.figma.com/file/fKyLulu6MUjQzQa9pBFFrC/Insights-and-Analytics?node-id=586%3A117159) and you can see the threads widget and modal
> > 
> > 
> > @BenCookie95 , I've updated response to have the things you mentioned.
> 
> Looks great aside from some things I noticed while adding the frontend:
> 
> 1. I noticed that there will also be instances where the message might be a file or an image. This means that we would have to include the metadata in the response
> 2. The response still includes threads even if the parent message is deleted
> 3. The response payload is missing the author details
> 
> @mkraft Do you think it's possible for us to handle file/image attachments? Or will there be too many joins?

@BenCookie95 , while I'll wait for your suggestions on
 - file attachments in the response
 - user displayname(nickname?)

Here are some things that got pushed in the last commit.
 - full post object is under `post` field.
 - deleted root posts won't be in the top threads anymore.

And a clarification is
 - The user details (only fields lastpictureupdate, firstname, lastname, id) would be in `user_information`. They work in storetests too, please let me know if you still don't see it. I'll look better for what could be going wrong.

plant99 commented:
Hey @BenCookie95 , sorry would need some clarification.
For file attachments `response.post.file_ids` is to be referred.

I could find a `fileinfo` table. If you'd want, I can expand the `file_ids` to contain `fileinfo` row.

Please also let me know if I'm missing other things, (if other attachments are stored in some other table), and I'll append them to response?

plant99 commented:
> That example is from when I sent an image, take a look at the server code and see how we include metadata in the post object for links/images/files. The relevant tables are `fileinfo` and `linkmetadata`. Let me know if you need some help
Thanks Ben, this is clear to me. I'll update tomorrow.

mkraft commented on store/storetest/thread_store.go:
```diff
@@ -1226,3 +1227,173 @@ func testMarkAllAsReadByChannels(t *testing.T, ss store.Store) {
+		require.Equal(t, topThreadsByUser2.Items[1].Post.Id, post3.Id)
+	})
+
+}
```
There are some cases that we should include here (I may have missed them if they're there):
1. results don't include threads from a different team
2. results don't include private channels a user is not a member of
3. results don't include DM threads
4. results only include threads within the given time frame
5. something testing the pagination
If these cases (and any others you can think of) are covered in the API tests that's totally fine. We just want to ensure that all of the logical cases are covered.

plant99 replied:
This PR only had test case 2 covered.

But [this commit](https://github.com/mattermost/mattermost-server/pull/20195/commits/07985a3dda589769fae3c09b84d895ca9cd32a60) adds the other test cases you mentioned.

plant99 commented on api4/insights.go:
```diff
@@ -202,3 +206,93 @@ func getTopChannelsForUserSince(c *Context, w http.ResponseWriter, r *http.Reque
+	c.Params.TeamId = r.URL.Query().Get("team_id")
+
+	// TeamId is an optional parameter
+	if c.Params.TeamId != "" {
```
https://github.com/mattermost/mattermost-server/pull/19953/files#diff-91cef20bc0ee91cdbef3d89f0495ce55944452ff8fd442644fa815a9261e8b37R70 didn't have a lot of design context, so did what was done in top-channels. Please let me know if I should enforce TeamId constraint even for user-threads.

mkraft replied:
Make sense. Thanks.

plant99 commented on model/insights.go:
```diff
@@ -50,6 +50,32 @@ type TopChannel struct {
+	Name            string                      `json:"channel_name"`
+	Message         string                      `json:"message"`
+	Participants    StringArray                 `json:"participants"`
+	UserId          string                      `json:"user_id"`
```
As discussed,
1) we have to separate the DB output and Response model if we mind the extra `UserId` here. Similar to PostId, this helps to populate UserInformation later, just like PostId populates `Post`.
2) We can use `Query` instead of `Select` but that's sparsely used in the repository.

plant99 commented on api4/insights.go:
```diff
@@ -202,3 +206,93 @@ func getTopChannelsForUserSince(c *Context, w http.ResponseWriter, r *http.Reque
+		c.SetPermissionError(model.PermissionViewTeam)
+		return
+	}
+
```
Done, copied from the usergroups API. https://github.com/mattermost/mattermost-server/pull/20195/commits/1cc82478d3f2a66079bc8fa73ade3adf19b28a5c

plant99 commented on store/sqlstore/thread_store.go:
```diff
@@ -935,3 +935,197 @@ func (s *SqlThreadStore) GetThreadUnreadReplyCount(threadMembership *model.Threa
+	for _, topThread := range topThreads {
+		userIDs := []string{topThread.UserId}
+		var err error
+		users, err := s.User().GetProfileByIds(context.Background(), userIDs, &store.UserGetByIdsOpts{}, true)
```
Done, thanks for catching this!

BenCookie95 commented:
> > That example is from when I sent an image, take a look at the server code and see how we include metadata in the post object for links/images/files. The relevant tables are `fileinfo` and `linkmetadata`. Let me know if you need some help
> 
> Thanks Ben, this is clear to me. I'll update tomorrow.

Hey @plant99, can we also add `nickname` and `username` to the user_information as well please?

BenCookie95 commented:
> Hey @BenCookie95, added the fields.

Perfect thanks! Last thing for me is that we need to just cleanup the response object, we can remove the following fields because they already exist in the `user_information` and `post` objects:
post_id
reply_count
message
user_id

plant99 commented:
> Perfect thanks! Last thing for me is that we need to just cleanup the response object, we can remove the following fields because they already exist in the `user_information` and `post` objects: post_id reply_count message user_id

@BenCookie95 please check, i've removed these fields.

BenCookie95 commented:
/update-branch

furqanmlk commented:
> https://mattermost-server-pr-20195.test.mattermost.cloud
 
Hi @BenCookie95 
This test server is giving BAD GATEWAY error.
I will update once it is up or I will create new one.

<img width="1059" alt="image" src="https://user-images.githubusercontent.com/6909890/172707790-72731892-f718-4677-b3e0-5459023575e1.png">

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/206742

furqanmlk reviewed (commented):
@plant99 Tested `today`, `7_day` and `28_day` in `Team insight` and `My insight` Working as expected.
One question:
If a root post is 25 days old and last reply was made today, that thread will be returned by time `range=28_day` or `today`?

plant99 commented:
> @plant99 Tested `today`, `7_day` and `28_day` in `Team insight` and `My insight` Working as expected. One question: If a root post is 25 days old and last reply was made today, that thread will be returned by time `range=28_day` or `today`?

It'll also come under today since the filter is on `thread.LastReplyAt`
